### DISCUSSION TITLE
v24.12.20

### DISCUSSION THREAD

**Discussion by davidrapan** [davidrapan/ha-solarman#337] [Announcements]
## What's Changed
* Add support for **time synchronization** in cloudless environments for inverters which utilizes heartbeat packets (e.g. Deye Microinverters)
* [feat: Gen/SL/MI sensors named by selected IO mode](https://github.com/davidrapan/ha-solarman/commit/7e94a4d6b8bceb11e44153faa013d83eaae42ff1)
  * Deye: Generator | SmartLoad | Microinverter sensors named according to IO Mode selected
* [feat: Add Today's and Total Generation - Deye](https://github.com/davidrapan/ha-solarman/commit/a08ed6fa56502ff3609ca033809f36502b805536)
* chore: Add tested firmware version - Deye Microinverter by @CrazyUs3r in https://github.com/davidrapan/ha-solarman/pull/330

## Experimental
* Battery Capacity sensor for Deye Hybrid Inverters
  * Estimates capacity of the battery during (dis)charging

**Full Changelog**: https://github.com/davidrapan/ha-solarman/compare/v24.12.16...v24.12.20

<hr /><em>This discussion was created from the release <a href='https://github.com/davidrapan/ha-solarman/releases/tag/v24.12.20'>v24.12.20</a>.</em>

**Comment by sofkaski** [comment]
I installed `v24.12.20_3` to my environment and did some silk glove testing. Notes:

Generally the release seems to work and is compatible with the previous release

Time-of-use behavior compared to what I reported in [v24.12.16 discussion](https://github.com/davidrapan/ha-solarman/discussions/327#discussioncomment-11595493) has changed. Now ticking days Monday-Thursday and Saturday (i.e. leaving out Friday and Sunday) results in `Enabled` state.  (These are still the same:  all days selected ->  `Week` and Sunday off -> `Week-Sunday` )

I got the following log message about 5 hours after the restart: 

> `2024-12-21 15:35:08.807 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/config/custom_components/solarman/api.py", line 70, in _heartbeat_response
    v5_frame[5] = v5_frame[5] + 1
    ~~~~~~~~^^^
ValueError: byte must be in range(0, 256)

Should the the counter roll-over at 255? ( `v5_frame[5] = (v5_frame[5] + 1) % 255` ?)

I'll look into the experimental `sensor.inverter_battery_capacity` once I get my battery to do one full charge-discharge cycle. I can compare this to the value my  BMS (Seplos V2) reports. I have it integrated to HA also.

**Reply by sofkaski** [reply]
Testing the new feature resulting in `sensor.inverter_today_generation` gives me pure zero since I have about 30 centimeters snow on my panels . :disappointed:

**Reply by davidrapan** [reply]
`sensor.inverter_today_generation` is just from generator port.

> Should the counter roll-over at 255?

Yeah but proly mod 256. Do a PR. 😉

Edit2: `(v5_frame[5] + 1) & 0xFF` will be better I guess.

**Reply by sofkaski** [reply]
> `sensor.inverter_today_generation` is just from generator port.
> 
Ah..yes. I  thought only from narrow perspective of solar panels :grin: .

**Reply by sofkaski** [reply]
> ... Do a PR. 😉
> 
Too late. I was about to do it now in the morning, but you had fixed it already.

**Reply by davidrapan** [reply]
Yeah I was moving some changes to the pysolarman from the repo.

**Comment by sofkaski** [comment]
I installed 8381431550. After a few hours I can probably confirm the fix.

**Reply by sofkaski** [reply]
Seems that fix  f7555b52513 killed the bug. Integration has been running enough long time so that the rollover had happened.

**Reply by davidrapan** [reply]
The code is now in submodule but great! Thanks for the feedback!
